src/main/java/org/apache/maven/plugins/deploy/DeployMojo.java

```diff
@@ -170,12 +183,71 @@ public void execute()
         {
             synchronized ( DEPLOYREQUESTS )
             {
-                while ( !DEPLOYREQUESTS.isEmpty() )
+                int requests = DEPLOYREQUESTS.size();
+                CompletionService<String> service = null;
+                if ( threads > 1 )
                 {
-                    ArtifactRepository repo = getDeploymentRepository( DEPLOYREQUESTS.get( 0 ) );
+                    service = new ExecutorCompletionService( new ScheduledThreadPoolExecutor( threads ) );
+                    getLog().info( "Deploying with " + threads + " threads" );
+                }
 
-                    deployProject( getSession().getProjectBuildingRequest(), DEPLOYREQUESTS.remove( 0 ), repo );
+                for ( final ProjectDeployerRequest projectDeployerRequest : DEPLOYREQUESTS )
+                {
+                    Callable<String> callable = new Callable<String>()
+                    {
+                        @Override
+                        public String call() throws Exception
+                        {
+                            ArtifactRepository repo = getDeploymentRepository( projectDeployerRequest );
+                            deployProject( getSession().getProjectBuildingRequest(), projectDeployerRequest, repo );
+                            return projectDeployerRequest.getProject().getName();
+                        }
+                    };
+
+                    if ( threads > 1 )
+                    {
+                        getLog().info( "Submitting " + projectDeployerRequest.getProject().getName() );
+                        service.submit( callable );
+                    }
+                    else
```

Comment:
The `else` could be avoided by using an executor with just a single worker thread.

Reply:
I did this to maintain passivity by leaving the upload on the current thread.

Reply:
You could do that by using a ExecutionService which runs the workers in the same thread like the DirectExecutor from https://google.github.io/guava/releases/24.0-jre/api/docs/index.html?com/google/common/util/concurrent/MoreExecutors.html

Reply:
please no dependency on guava....

Reply:
I'd create a similar service in Maven's support libraries. Being able to switch freely between parallel and serial execution with the same code is probably useful elsewhere (like running unit tests or reports).